src/compressor/QatAccel.cc

```diff
@@ -11,67 +11,100 @@
+    qzClose(session);
+  }
+  delete session;
+}
+
 /* Estimate data expansion after decompression */
 static const unsigned int expansion_ratio[] = {5, 20, 50, 100, 200};
 
-QatAccel::~QatAccel() {
-  if (NULL != session.internal) {
-    qzTeardownSession(&session);
-    qzClose(&session);
+QatAccel::QatAccel() {
+  sessions.resize(SESSIONS_NUM);
+  for (auto &session : sessions) {
+    session.reset(new struct QzSession_S);
+    memset(session.get(), 0, sizeof(struct QzSession_S));
   }
 }
 
+QatAccel::~QatAccel() {}
+
 bool QatAccel::init(const std::string &alg) {
   QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
   int rc;
 
+  std::unique_lock<std::mutex> l(lock);
+  if (!alg.empty()) {
+    return true;
+  }
+
   rc = qzGetDefaults(&params);
   if (rc != QZ_OK)
     return false;
   params.direction = QZ_DIR_BOTH;
-  if (alg == "snappy")
-    params.comp_algorithm = QZ_SNAPPY;
-  else if (alg == "zlib")
+  if (alg == "zlib")
     params.comp_algorithm = QZ_DEFLATE;
-  else if (alg == "lz4")
-    params.comp_algorithm = QZ_LZ4;
   else
     return false;
 
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
 
-  rc = qzInit(&session, QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
-    return false;
+  for (auto &session : sessions) {
+    rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
+    if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+      return false;
 
-  rc = qzSetupSession(&session, &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(&session);
-    qzClose(&session);
-    return false;
+    rc = qzSetupSession(session.get(), &params);
+    if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
+      qzTeardownSession(session.get());
+      qzClose(session.get());
+      return false;
+    }
   }
 
+  this->alg = alg;
   return true;
 }
 
 int QatAccel::compress(const bufferlist &in, bufferlist &out, boost::optional<int32_t> &compressor_message) {
+  unique_session_ptr session;
+  {
+    std::unique_lock<std::mutex> l(lock);
+    if (sessions.empty()) {
+      conds.wait(l, [this](){return !sessions.empty();});
+    }
+    session = std::move(sessions.back());
+    sessions.pop_back();
+  }
+
   for (auto &i : in.buffers()) {
     const unsigned char* c_in = (unsigned char*) i.c_str();
     unsigned int len = i.length();
-    unsigned int out_len = qzMaxCompressedLength(len);
+    unsigned int out_len = qzMaxCompressedLength(len, session.get());
 
     bufferptr ptr = buffer::create_small_page_aligned(out_len);
-    int rc = qzCompress(&session, c_in, &len, (unsigned char *)ptr.c_str(), &out_len, 1);
+    int rc = qzCompress(session.get(), c_in, &len, (unsigned char *)ptr.c_str(), &out_len, 1);
     if (rc != QZ_OK)
       return -1;
     out.append(ptr, 0, out_len);
   }
 
+  {
+    std::unique_lock<std::mutex> l(lock);
+    sessions.push_back(std::move(session));
+    conds.notify_one();
+  }
```

Comment:
nit, might want to use `scope_guard` to release the session. so the resource acquiring and releasing could be colocated in a single place for better readability.

Reply:
or better off:
```c++
// put the session back to the session pool in a RAII manner
struct cached_session_t {
  cached_session_t(QatAccel* accel) 
    : accel{accel} {
    std::unique_lock lock{accel->mutex};
    conds.wait(lock, [accel] { return !accel->sessions.empty(); });
    session = std::move(sessions.back());
    accel->sessions.pop_back();
  }
  ~cached_session_t() {
    std::unique_lock lock{accel->mutex};
    accel->sessions.push_back(std::move(session));
    accel->conds.notify_one();
  }
  QatAccel* accel;
  QatAccel::unique_session_ptr session;
};

// ...
class QatAccel {
  // ...
  friend struct cached_session_t;
};
```

Reply:
[OT], might want to rename `unique_session_ptr` to `session_ptr`.

Reply:
I use `struct cached_session_t` in code. It makes the code concise。
I has renamed `unique_session_ptr` to `session_ptr`.

Reply:
> nit, might want to use `scope_guard` to release the session. so the resource acquiring and releasing could be colocated in a single place for better readability.

Hi, @tchaikov , I used `scope_lock` to release the session. Is it right?
```
std::scoped_lock lock{accel->mutex};
```

Reply:
if the proposed cached_session change is incorporated, the scoped _guard suggestion can be ignored. It was proposed as an alternative of the cached_session solution.

Reply:
> if the proposed cached_session change is incorporated, the scoped _guard suggestion can be ignored. It was proposed as an alternative of the cached_session solution.

OK, I have used `cached_session_t`. Thanks.